WooCommerce/Classes/ViewRelated/Orders/Collect Payments/CollectOrderPaymentUseCase.swift

```diff
@@ -267,7 +267,8 @@ private extension CollectOrderPaymentUseCase {
             }
 
             guard self.isOrderAwaitingPayment() else {
-                return onCheckCompletion(.failure(CollectOrderPaymentUseCaseError.orderAlreadyPaid))
+                return onPaymentCompletion(.success(CardPresentCapturedPaymentData(paymentMethod: .unknown,
```

Comment:
I don't think we should return success here, in the existing flows. The error is clearer for some of the situations which can come up.

For POS, where we're really confident that the order's just been created, we know that the most likely (perhaps only possible) reason for hitting this error is after retry after the missed network connection as covered in #13165. In that case, the shopper will have just tapped their card, so it's not confusing to see a success payment.

For the existing flow, where we rely more on storage and orders can be much older, it's not so safe to show success here. A merchant could have an outdated copy of the order from storage, tap `Collect Payment`, then immediately see that the order's complete without the customer tapping their card, nor any explanation of how it's complete.

In that situation, they're likely to think that the app is wrong and that they need to redo the order. The error would have explained it much more clearly.

The best place to do this for the new experience is probably in `CardPresentPaymentAlertPresenterAdaptor.present`, e.g. with a case like this one:

```
    case .paymentError(error: CollectOrderPaymentUseCaseError.orderAlreadyPaid, _, _),
                .paymentErrorNonRetryable(error: CollectOrderPaymentUseCaseError.orderAlreadyPaid, _):
            paymentEventSubject.send(.show(eventDetails: .paymentSuccess(done: {})))
```

It probably needs to do more than _just_ that, e.g. we'll need to clean up something in the facade, but that's my idea for a starting point... although from a quick try, it does look like it "just works".

Reply:
> I don't think we should return success here, in the existing flows. The error is clearer for some of the situations which can come up.
> 
> For POS, where we're really confident that the order's just been created, we know that the most likely (perhaps only possible) reason for hitting this error is after retry after the missed network connection as covered in #13165. In that case, the shopper will have just tapped their card, so it's not confusing to see a success payment.
> 
> For the existing flow, where we rely more on storage and orders can be much older, it's not so safe to show success here. A merchant could have an outdated copy of the order from storage, tap `Collect Payment`, then immediately see that the order's complete without the customer tapping their card, nor any explanation of how it's complete.
> 
> In that situation, they're likely to think that the app is wrong and that they need to redo the order. The error would have explained it much more clearly.

Thanks a lot for the explanation about why this change shouldn't be applied to the existing flow, it makes sense. @staskus and I had a pair programming session to work on this issue, and your suggested fix works from my testing! (I tested that this scenario results in success for POS and remains an error in the existing flow in order creation).

> It probably needs to do more than just that, e.g. we'll need to clean up something in the facade, but that's my idea for a starting point... although from a quick try, it does look like it "just works".

For this, I'd like to check with you. There are also [2](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L24) [other](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L33) TODO comments about calling the facade's `cancelPayment` for further cleanup. I tried triggering `CardPresentPaymentService.cancelPayment` in the event's `cancelPayment` closure, but the current implementation of cancellation is to also call the event's `cancelPayment` closure:

https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift#L144-L146

In `CardPresentPaymentCollectOrderPaymentUseCaseAdaptor`, we could also change how we handle the cancellation. An option is to cancel the payment intent, which I think we could do for non-retryable payment errors (there could be exceptions that I didn't know of). Does this sound like the right way to go, canceling the payment intent for non-retryable payment errors (moving [this line](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentCollectOrderPaymentUseCaseAdaptor.swift#L145) to the .processing cases below)?

For retryable payment errors, is there any cleanup we should do? In place of [this TODO comment](https://github.com/woocommerce/woocommerce-ios/blob/802b6bb208d62810f11868ab328051a30ab27130/WooCommerce/Classes/POS/Card%20Present%20Payments/CardPresentPaymentsAlertPresenterAdaptor.swift#L24).

If this is a bit too much to take on before your AFK, I can create a separate task for this further cancelation cleanup for later.

Reply:
For this case I don't think it would be appropriate or needed to call `cancelPayment`. The `orderAlreadyPaid` error is thrown before we create the payment intent, so there's nothing to cancel yet.

From our testing, it's working without further cleanup, so that's great. It makes sense actually, because we throw away the `CollectOrderPaymentUseCase` when the merchant taps `New transaction`.

However, `CollectOrderPaymentUseCase` doesn't know anything about this `success`, so we're tracking this as a payment failure in our analytics... or actually two failures:

![CleanShot 2024-06-28 at 12 06 37@2x](https://github.com/woocommerce/woocommerce-ios/assets/2472348/dac8415a-86c8-4079-9550-18d6d3da0e15)

We're not really dealing with this yet, but it would be nice to track success if we _can_. This can be delayed but let's capture it in a ticket. It's fine to track the failures, but we should track the success as well at the end.

I think perhaps we should defer those TODOs for after my AFK – we'll have a new project to tighten everything up and deal with the TODOs... I can't dig in to them now.

Reply:
> However, `CollectOrderPaymentUseCase` doesn't know anything about this `success`, so we're tracking this as a payment failure in our analytics... or actually two failures:
> 
> ![CleanShot 2024-06-28 at 12 06 37@2x](https://private-user-images.githubusercontent.com/2472348/344140191-dac8415a-86c8-4079-9550-18d6d3da0e15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1ODEyNjksIm5iZiI6MTcxOTU4MDk2OSwicGF0aCI6Ii8yNDcyMzQ4LzM0NDE0MDE5MS1kYWM4NDE1YS04NmM4LTQwNzktOTU1MC0xOGQ2ZDNkYTBlMTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjhUMTMyMjQ5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjUwYmJiZjRkZTIxMmNlZDMzNWRjMzMzNGNiNmIxZDVhNTQxZDM5NjNkYjNiOTQ3NTdkNjAzMzZkMDAzZWUzYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.U7I-cFD-ZhKJP21y4tEpCBeukvKIQht5BjJCi_sJGVg)
> 
> We're not really dealing with this yet, but it would be nice to track success if we _can_. This can be delayed but let's capture it in a ticket. It's fine to track the failures, but we should track the success as well at the end.

Good catch on the tracking side, yea we'd want to make sure to track the success state for this case. I created a [draft task](https://github.com/orgs/woocommerce/projects/247/views/6?sliceBy%5Bvalue%5D=iOS+%F0%9F%8D%8E&pane=issue&itemId=69143334) for this, currently in the "Blocked" column because we don't have an analytics plan yet. I suggested tracking POS success when the view model receives the payment success event and is about to show the success UI.